src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -767,14 +766,28 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
     }
 
     private fun checkAttr(holder: RsAnnotationHolder, attr: RsAttr) {
-        checkImplBothCopyAndDrop(holder, attr)
+        checkDeriveAttribute(holder, attr)
         checkInlineAttr(holder, attr)
         checkReprAttribute(holder, attr)
 
         if (attr.owner !is RsFunction)
             checkStartAttribute(holder, attr)
     }
 
+    private fun checkDeriveAttribute(holder: RsAnnotationHolder, attr: RsAttr) {
+        if (attr.metaItem.name != "derive") return
+
+        when (attr.owner) {
+            is RsStructItem, is RsEnumItem -> {
+                checkImplBothCopyAndDrop(holder, attr)
+            }
+
+            else -> {
+                RsDiagnostic.DeriveAttrUnsupportedItem(attr.metaItem, "`derive` may only be applied to structs, enums and unions").addToHolder(holder)
```

Comment:
Let's move error message into diagnostic itself because I don't know how it can be customizable